Add a guide for building a distroless cargo container (#80)
SUSE/bci-docs: content/guides/building-a-distroless-image.adoc

---
Title: How to build a distroless image using SLE BCI
---

== What is a distroless image?

Distroless images are stripped down container images with nearly nothing left in
them from a distribution, hence the name distroless. Usually they only contain
certificates and maybe some core libraries, but no shell or other core utilities
like `cat` or even `ls`.

These images have the advantage of potentially fewer vulnerabilities, as there
is less software included as well as being overall smaller. The major
disadvantage is that debugging the containerized application is very cumbersome,
as the container image includes no debugging tools whatsoever and even lacks
tools to read log files. Debugging applications usually requires attaching
sidecar containers or interacting with the container via the `/proc/`
filesystem.

> 

Reviewer: I suggest to add something like: Using a distroless approach for multiple use cases or applications will make it harder for having the same base standard for your images. Thus it will potentially increase your maintenance and administration of these images. 
Therefore we recommend to checkout our SLE BCI Micro or SLE BCI Busybox images first.

Author: That's not really valid, if you want to deploy an application as a container, then you have to maintain the build irrespective of the base or target image. Whether that is `SCRATCH`, SLE BCI Micro or something entirely different doesn't really make a huge difference.

Reviewer: ok got it, I was just thinking about how to make people check Micro and Busybox before starting from scratch and do "distroless". I guess it's hard without leaving the technical side and jumping into the marketing side...

Author: I have added a warning that points out a big catch with `FROM SCRATCH`: your rpmdb is missing. If you use a security scanner that uses only that to find vulnerabilities, then it'll find none :-D

---
Title: How to build a distroless image using SLE BCI
---

== What is a distroless image?

Distroless images are stripped down container images, where the underlying Linux
distribution is reduced to the bare minimum. A distroless image normally
contains only certificates and specific core libraries, and it does not include
a shell or utilities like `cat` or `ls`.

The advantages of distroless images include smaller size and potentially fewer
vulnerabilities.

The major disadvantage is the difficulty of debugging a containerized
application, as the container image does not provide any debugging tools and may
even lack tools to read log files. Debugging applications usually requires
attaching sidecar containers or interacting with the container via the `/proc/`
filesystem. For this reson, we recommend to use SLE BCI Base, SLE BCI Micro or
SLE BCI BusyBox as the deployment image, because they make debugging easier, and
they come with a valid rpm database.

{{< hint type="warning" >}}
A distroless image, or any image without the rpm database, makes it harder for

> container security scanners to find known vulnerabilities in the container

Reviewer 2: That's only half the truth though. If you generate and manage a SBOM for the container you could still have sufficient information to find out about vulnerabilities inside.

Author: I was assuming that someone who manages an SBOM for their container image would know this. Also, the text is explicitly saying "makes it harder", not "makes it impossible".

detect a vulnerable shared library.
{{< /hint >}}

=== How to build a distroless images from SLE BCI

Although SUSE does not offer a distroless SLE BCI, you can build one by creating
a multi-stage build and creating a final image based on `SCRATCH`. The following
tutorial uses an existing application to demonstrate how to identify its
dependent libraries and copy them into the final image.

The first step is to identify all the components required by the
application. This can include configuration files, external binaries, and shared
libraries. It is your task to determine which configuration files and binaries
are required for the program to function correctly. For example, a Python
application will require that the Python interpreter is present in final image.

As minimum, compiled applications are usually dynamically linked to libc. It is
possible to statically link against most libraries, but not against glibc (which
is the libc implementation used by SLE). Therefore the compiled application
requires at least the shared libc library. The required libraries can be
identified using `ldd`. The following example deploys the Rust package manager

> `cargo` in an empty image.

Reviewer 2: Why did you pick cargo for this example? Wouldn't it make more sense to demonstrate how to build a rust *application* using BCI's cargo (providing all the security benefits of the certified maintained SLE etc) and then give the instructions on how to make a distroless container or of the application?

Reviewer 2: The way it is now is that it implies there would be a benefit of using this distroless cargo in the rust bulling process which I don't think is the case.

Author: The focus of this guide was not how to build an application, but how to identify the required shared libraries. `cargo` was just the first example that came to my mind. I don't see where this guide implies any benefits from using a containerized cargo. It's imho actually rather silly to containerize cargo, but then again, that was really not the focus of this exercise.

---
Title: How to build a distroless image using SLE BCI
---

== What is a distroless image?

> Distroless images are stripped down container images with nearly nothing left in
> them from a distribution, hence the name distroless. Usually they only contain
> certificates and maybe some core libraries, but no shell or other core utilities
> like `cat` or even `ls`.

Reviewer 3: ```suggestion
Distroless images are stripped down container images, where the underlying Linux distribution is reduced to the bare minimum. A distroless image normally contains only certificates and specific core libraries, and it does not include shell or utilities like `cat` or `ls`.
```

> These images have the advantage of potentially fewer vulnerabilities, as there
> is less software included as well as being overall smaller. The major
> disadvantage is that debugging the containerized application is very cumbersome,
> as the container image includes no debugging tools whatsoever and even lacks
> tools to read log files. Debugging applications usually requires attaching
> sidecar containers or interacting with the container via the `/proc/`
> filesystem. We recommend to use SLE BCI Base, SLE BCI Micro or SLE BCI BusyBox
> as the deployment image, due to the much better debugging UX and because they
> still contain a valid rpm database.
> 

Reviewer 3: ```suggestion
The advantages of distroless images include smaller size and potentially fewer vulnerabilities.

The major disadvantage is the difficulty of debugging a containerized application,
as the container image does not provide any debugging tools and may even lack
tools to read log files. Debugging applications usually requires attaching
sidecar containers or interacting with the container via the `/proc/`
filesystem. For this reson, we recommend to use SLE BCI Base, SLE BCI Micro or SLE BCI BusyBox
as the deployment image, because they make debugging easier, and they come with a valid rpm database.

```

{{< hint type="warning" >}}

> A distroless image or any image without the rpm database will make it harder for
> container security scanners to find known vulnerabilities in your container
> image. Exercise caution when a scanner reports no vulnerabilities in your
> distroless image, as it might be solely relying on the rpm database and thus is
> unable to detect vulnerable shared libraries.

Reviewer 3: ```suggestion
A distroless image, or any image without the rpm database, makes it harder for
container security scanners to find known vulnerabilities in the container
image. Keep in mind that if a scanner relies only on the rpm database, it cannot detect vulnerable shared library.
```

{{< /hint >}}

=== How to build a distroless images from SLE BCI

> SUSE does not offer a distroless SLE BCI instead you can build one yourself by
> creating a multistage build and creating a final image based on `SCRATCH`. For
> this tutorial, we will use an existing application, identify its dependent
> libraries and copy them into our final image.

Reviewer 3: ```suggestion
Although SUSE does not offer a distroless SLE BCI, you can build one by
creating a multi-stage build and creating a final image based on `SCRATCH`. The following tutorial uses an existing application to demonstrate how to identify its dependent libraries and copy them into the final image.
```

> The first step of creating a distroless image is to identify all the components
> required by your application. This can include configuration files, external
> binaries and shared libraries. You as the author of your program have to figure
> out yourself which configuration files and binaries are required for your
> program to function correctly. For example a python application will require
> that the python interpreter is present in final image.

Reviewer 3: ```suggestion
The first step is to identify all the components
required by the application. This can include configuration files, external
binaries, and shared libraries. It is your task to determine which configuration files and binaries are required for the program to function correctly. For example, a Python application may require
that the Python interpreter is present in final image.
```

> Compiled applications are usually dynamically linked to at least libc. It is
> possible to statically link against most libraries, but not against glibc (the
> libc implementation used by most Linux distributions, including SLE). Therefore
> our compiled application requires at least the shared libc library. We can
> identify the required libraries using `ldd` and will use Rust's package manager
> `cargo` as the example application to deploy in a mostly empty image.

Reviewer 3: ```suggestion
As minimum, compiled applications are usually dynamically linked to libc. It is
possible to statically link against most libraries, but not against glibc (which is the
libc implementation used by SLE). Therefore the compiled application requires at least the shared libc library. The required libraries can be identified using `ldd`. The following example deploys the Rust package manager `cargo` in an empty image.
```

        liblber-2.4.so.2 => /usr/lib64/liblber-2.4.so.2 (0x00007f3764bd5000)
        libzstd.so.1 => /usr/lib64/libzstd.so.1 (0x00007f37648a5000)
        libbrotlidec.so.1 => /usr/lib64/libbrotlidec.so.1 (0x00007f3764699000)
        libjitterentropy.so.3 => /usr/lib64/libjitterentropy.so.3 (0x00007f3764492000)
        libunistring.so.2 => /usr/lib64/libunistring.so.2 (0x00007f376410f000)
        libkrb5.so.3 => /usr/lib64/libkrb5.so.3 (0x00007f3763e36000)
        libk5crypto.so.3 => /usr/lib64/libk5crypto.so.3 (0x00007f3763c1e000)
        libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007f3763a1a000)
        libkrb5support.so.0 => /usr/lib64/libkrb5support.so.0 (0x00007f376380b000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f37635f3000)
        libsasl2.so.3 => /usr/lib64/libsasl2.so.3 (0x00007f37633d6000)
        libbrotlicommon.so.1 => /usr/lib64/libbrotlicommon.so.1 (0x00007f37631b5000)
        libkeyutils.so.1 => /usr/lib64/libkeyutils.so.1 (0x00007f3762fb0000)
        libselinux.so.1 => /lib64/libselinux.so.1 (0x00007f3762d87000)
        libpcre.so.1 => /usr/lib64/libpcre.so.1 (0x00007f3762afe000)

----

> The above output are all shared libraries that we will have to copy into the
> final image, with one exception: `linux-vdso.so.1`. This shared library is not
> present on the file system, it is a virtual shared library exported by the
> kernel for improved performance. Consult the manpage of `vdso` via `man vdso`
> for more information about `linux-vdso.so.1`.

Reviewer 3: ```suggestion
The output lists all shared libraries that must be copied into the
final image, with one exception: `linux-vdso.so.1`. This shared library is not
present on the file system, it is a virtual shared library exported by the
kernel for improved performance. Consult the manpage of `vdso` via `man vdso`
for more information about `linux-vdso.so.1`.
```

The remaining shared libraries are required. We will parse the output of `ldd`
and copy them into a directory including their full paths to then copy them into

> the final distroless image.

Reviewer 4: I think he meant creating the entire directory structure for the shared libraries. For e.g. for `libcurl`, the whole tree should be instantiated (directories should be created): `/usr/lib64/libcurl.so.4`

We first handle the shared libraries which are linked just by their name and not
their full path as follows:
[source,bash]
----
for lib in $(ldd /usr/bin/cargo | awk '{print $3}'); do
    install -Dp $lib /l$lib
done
----

> The above snippet will copy all libraries including their directory structure
> into the directory `/l/`:

Reviewer 3: ```suggestion
The code above copies all libraries, including their directory structure,
into the directory `/l/`:
```

[source,Shell]
----
# tree /l
/l
|-- lib64
|   |-- libc.so.6
|   |-- libcom_err.so.2
|   |-- libdl.so.2
|   |-- libgcc_s.so.1
|   |-- libm.so.6
|   |-- libpthread.so.0
|   |-- libresolv.so.2
|   |-- libselinux.so.1
|   `-- libz.so.1
`-- usr
    `-- lib64
        |-- libbrotlicommon.so.1
        |-- libbrotlidec.so.1
        |-- libcrypto.so.1.1
        |-- libcurl.so.4
        |-- libgssapi_krb5.so.2
        |-- libidn2.so.0
        |-- libjitterentropy.so.3
        |-- libk5crypto.so.3
        |-- libkeyutils.so.1
        |-- libkrb5.so.3
        |-- libkrb5support.so.0
        |-- liblber-2.4.so.2
        |-- libldap_r-2.4.so.2
        |-- libnghttp2.so.14
        |-- libpcre.so.1
        |-- libpsl.so.5
        |-- libsasl2.so.3
        |-- libssh.so.4
        |-- libssl.so.1.1
        |-- libunistring.so.2
        `-- libzstd.so.1

3 directories, 30 files

----

We are still missing the shared libraries linked by their full path, in this

> case that is only `/lib64/ld-linux-x86-64.so.2`. We perform the same using the
> following snippet:

Reviewer 4: ```suggestion
case that is only `/lib64/ld-linux-x86-64.so.2`. We can copy them using the
following snippet:
```

> following snippet:

Author: Better now? It was about the copy.

[source,bash]
----

> for lib in $(ldd /usr/bin/cargo |sed '/=>/d'|grep "^[[:space:]]*/"|awk '{print $1}'); do

Reviewer 4: Nice command-line fu. Couple of comments: 
1. Just curious, did you try this on a SLE machine? I just ran `podman run -it registry.suse.com/bci/bci-base:15.4 bash`(from the Containerfile below) and it didn't have `awk` installed.
2. Why do we need the sed and grep? 
```suggestion
for lib in $(ldd /usr/bin/cargo | awk '{print $1}'); do
```

Author: yes we need the grep:
```ShellSession
2a98563b1d21:/ # ldd /usr/bin/cargo |grep "^[[:space:]]*/"|cut -d" " -f 1
        /lib64/ld-linux-x86-64.so.2
2a98563b1d21:/ # ldd /usr/bin/cargo |cut -d" " -f 1
        linux-vdso.so.1
        libz.so.1
        libcurl.so.4
        libssl.so.1.1
        libcrypto.so.1.1
        libgcc_s.so.1
        libpthread.so.0
        libm.so.6
        libdl.so.2
        libc.so.6
        /lib64/ld-linux-x86-64.so.2
        libnghttp2.so.14
        libidn2.so.0
        libssh.so.4
        libpsl.so.5
        libgssapi_krb5.so.2
        libldap_r-2.4.so.2
        liblber-2.4.so.2
        libzstd.so.1
        libbrotlidec.so.1
        libjitterentropy.so.3
        libunistring.so.2
        libkrb5.so.3
        libk5crypto.so.3
        libcom_err.so.2
        libkrb5support.so.0
        libresolv.so.2
        libsasl2.so.3
        libbrotlicommon.so.1
        libkeyutils.so.1
        libselinux.so.1
        libpcre.so.1
```

    install -Dp $lib /l$lib
done
----

Thereby we obtain all necessary libraries under `/l/`:

[source,Shell]
----
# tree /l
/l
|-- lib64
|   |-- ld-linux-x86-64.so.2
|   |-- libc.so.6
|   |-- libcom_err.so.2
|   |-- libdl.so.2
|   |-- libgcc_s.so.1
|   |-- libm.so.6
|   |-- libpthread.so.0
|   |-- libresolv.so.2
|   |-- libselinux.so.1
|   `-- libz.so.1
`-- usr
    `-- lib64
        |-- libbrotlicommon.so.1
        |-- libbrotlidec.so.1
        |-- libcrypto.so.1.1
        |-- libcurl.so.4
        |-- libgssapi_krb5.so.2
        |-- libidn2.so.0
        |-- libjitterentropy.so.3
        |-- libk5crypto.so.3
        |-- libkeyutils.so.1
        |-- libkrb5.so.3
        |-- libkrb5support.so.0
        |-- liblber-2.4.so.2
        |-- libldap_r-2.4.so.2
        |-- libnghttp2.so.14
        |-- libpcre.so.1
        |-- libpsl.so.5
        |-- libsasl2.so.3
        |-- libssh.so.4
        |-- libssl.so.1.1
        |-- libunistring.so.2
        `-- libzstd.so.1

3 directories, 31 files
----

> We now combine our findings to start assembling our `Dockerfile`. We will use

Reviewer 4: `Dockerfile`: quite unrelated but since we're thinking of going ahead with podman as the container engine of choice, are we going to ever refer to `Containerfiles`?

Author: I personally would stick to `Dockerfile` now: all the runtimes support that and it's what most users are familiar with.

the SLE BCI Base image as our base image, install cargo and awk and create the
directory tree as illustrated above:

[source,Dockerfile]
----
FROM registry.suse.com/bci/bci-base:15.4 as builder

> RUN zypper -n in cargo awk
> 
> RUN for lib in $(ldd /usr/bin/cargo | awk '{print $3}'); do \
>         install -Dp $lib /l$lib; \
>      done
> RUN for lib in $(ldd /usr/bin/cargo |sed '/=>/d'|grep "^[[:space:]]*/"|awk '{print $1}'); do \
>         install -Dp $lib /l$lib; \
>      done
> ----

Reviewer 4: huge nit: but I think we can make do without installing awk
```suggestion
RUN zypper -n in cargo awk

RUN for lib in $(ldd /usr/bin/cargo |cut -d" " -f 3); do \
        install -Dp $lib /l$lib; \
     done
RUN for lib in $(ldd /usr/bin/cargo | cut -d" " -f 1); do \
        install -Dp $lib /l$lib; \
     done
----
```

Author: Nice idea, I've included something like that, but had to leave the `grep` in there, otherwise the second `RUN` will try to install a bunch of non-existent libraries.

In the second step we copy `cargo` itself and all libraries under `/l/` into an
empty image based on `SCRATCH`. As this image only contains `cargo`, we set both
`CMD` and `ENTRYPOINT` to cargo, so that the container will not behave in
unexpected ways when launched without parameters. We arrive at the following
complete `Dockerfile`:

[source,Dockerfile]
----
FROM registry.suse.com/bci/bci-base:15.4 as builder

RUN zypper -n in cargo awk

RUN for lib in $(ldd /usr/bin/cargo | awk '{print $3}'); do \
        install -Dp $lib /l$lib; \
     done
RUN for lib in $(ldd /usr/bin/cargo |sed '/=>/d'|grep "^[[:space:]]*/"|awk '{print $1}'); do \
        install -Dp $lib /l$lib; \
     done

FROM scratch

COPY --from=builder /l/ /
COPY --from=builder /usr/bin/cargo /usr/bin/cargo
CMD ["/usr/bin/cargo"]

> ENTRYPOINT ["/usr/bin/cargo"]

Reviewer 4: > we set both `CMD` and `ENTRYPOINT` to cargo, so that the container will not behave in unexpected ways when launched without parameters.

I don't think this does what it says it does, for instance, If I run this without any arguments, behind the scenes, it's actually executing `/usr/bin/cargo /usr/bin/cargo` and of course that's not right and it throws an error stating the same. I believe, just specifying `ENTRYPOINT` should be good enough since cargo, without any arguments would print help anyways. What do you think?

Author: You are of course right. Thanks!

----

We build this image with our favorite container runtime:

{{< tabs "build_cargo">}} {{< tab "Docker">}}

[source,Shell]
----
docker build -t cargo .
----

{{< /tab >}} {{< tab "Podman" >}}

[source,Shell]
----
buildah bud --layers -t cargo .
----

{{< /tab >}} {{< tab "nerdctl" >}}

[source,Shell]
----
nerdctl build -t cargo .
----

{{< /tab >}} {{< /tabs >}}

> We have successfully created a fully containerized cargo container, that we can
> now use:

Reviewer 3: ```suggestion
This creates a fully containerized ready-to-use cargo container:
```
